File: apps/nihongo_flutter/lib/data/remote/auth_api.dart
import 'package:dio/dio.dart';

import '../local/token_store.dart';

String? _extractAccessToken(dynamic data) {
  if (data is! Map) return null;
  final map = Map<String, dynamic>.from(data);
  final nested = map['data'];
  if (nested is Map && nested['access_token'] is String) {
    return nested['access_token'] as String;
  }
  final direct = map['access_token'];
  return direct is String ? direct : null;
}

class AuthApi {
  AuthApi(this._dio, this._tokenStore);
  final Dio _dio;
  final TokenStore _tokenStore;

  Future<void> login({required String email, required String password}) async {
    final res = await _dio.post<Map<String, dynamic>>(
      '/auth/login',
      data: {'email': email, 'password': password},
    );
    final token = _extractAccessToken(res.data);
    if (token == null || token.isEmpty) {
      throw DioException(
        requestOptions: res.requestOptions,
        message: 'Không nhận được access token',
      );
    }
    await _tokenStore.saveAccessToken(token);
  }

  Future<void> loginWithOidc({
    required String accessToken,
    String? idToken,
  }) async {
    final res = await _dio.post<Map<String, dynamic>>(
      '/auth/oidc',
      data: {
        'accessToken': accessToken,
        if (idToken != null && idToken.isNotEmpty) 'idToken': idToken,
      },
    );
    final token = _extractAccessToken(res.data);
    if (token == null || token.isEmpty) {
      throw DioException(
        requestOptions: res.requestOptions,
        message: 'Không nhận được access token từ OIDC',
      );
    }
    await _tokenStore.saveAccessToken(token);
  }

  Future<void> logout() async {
    try {
      await _dio.post('/auth/logout');
    } finally {
      await _tokenStore.clear();
    }
  }
}

File: apps/nihongo-ios/NihongoEDU/Data/Remote/AuthAPI.swift
import Foundation

struct AuthAPI {
    private let session: URLSession

    init(session: URLSession = .shared) {
        self.session = session
    }

    func login(email: String, password: String) async throws {
        try await postAuth(
            path: "/auth/login",
            body: ["email": email, "password": password]
        )
    }

    func loginOidc(accessToken: String, idToken: String?, refreshToken: String? = nil) async throws {
        var body: [String: Any] = ["accessToken": accessToken]
        if let idToken, !idToken.isEmpty { body["idToken"] = idToken }
        try await postAuth(path: "/auth/oidc", body: body)
        if let refreshToken, !refreshToken.isEmpty {
            TokenStore.saveRefreshToken(refreshToken)
        }
    }

    func logout() async {
        defer { TokenStore.clear() }
        guard let token = TokenStore.accessToken(),
              let url = URL(string: "\(APIConfig.baseURL)/auth/logout") else { return }
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        _ = try? await session.data(for: req)
    }

    private func postAuth(path: String, body: [String: Any]) async throws {
        guard let url = URL(string: "\(APIConfig.baseURL)\(path)") else {
            throw APIError.invalidURL
        }
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await session.data(for: req)
        if let http = response as? HTTPURLResponse, !(200..<300).contains(http.statusCode) {
            throw APIError.badStatus(http.statusCode)
        }

        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let token = Self.extractAccessToken(json) else {
            throw APIError.decodingFailed
        }
        TokenStore.saveAccessToken(token)
    }

    private static func extractAccessToken(_ json: [String: Any]) -> String? {
        if let nested = json["data"] as? [String: Any],
           let token = nested["access_token"] as? String, !token.isEmpty {
            return token
        }
        if let token = json["access_token"] as? String, !token.isEmpty {
            return token
        }
        return nil
    }
}
